torch/testing/_internal/common_utils.py

```diff
@@ -803,204 +810,412 @@ def is_integral(self, dtype):
         dtypes = [x for x in get_all_dtypes() if x not in get_all_complex_dtypes()]
         return dtype in dtypes and not dtype.is_floating_point
 
-    # accepts tensors, dtypes, or np.ndarrays
-    def get_default_tolerance(self, a, b=None):
-        if b is None:
-            dtype = torch.float
-            if isinstance(a, torch.Tensor):
-                dtype = a.dtype
-            elif isinstance(a, torch.dtype):
-                dtype = a
-            elif TEST_NUMPY and isinstance(a, numpy.ndarray):
-                # Some tests call assertEqual with numpy Unicode.
-                if numpy.issubdtype(a.dtype, numpy.dtype('U')):
-                    dtype = torch.float
-                else:
-                    dtype = torch.from_numpy(a).dtype
-            if self.is_integral(dtype):
-                return (0, 0)
-            dtype = str(dtype).split('.')[-1]
-            return self.dtype_precisions.get(dtype, (self.precision, self.precision))
-
-        a_tol = self.get_default_tolerance(a)
-        b_tol = self.get_default_tolerance(b)
-        return (max(a_tol[0], b_tol[0]), max(a_tol[1], b_tol[1]))
-
-    def assertEqual(self, x, y, message='', *, atol=None, rtol=None, allow_inf=False, exact_dtype=None):
+    # Returns a dtype suitable for comparing tensors a and b.
+    # This closely follows NumPy's type promotion logic, except it includes
+    # bfloat16 and accounts for the current state of PyTorch's bfloat16 and
+    # float16 implementations.
+    # Note: since torch.isclose converts its inputs to floating point values
+    # (it multiplies by rtol, a double) this function takes an optional
+    # allow_ints argument. If false, it only returns bool, float, or complex
+    # tensors. This can prevent a second conversion when comparing tensors.
+    def compare_type(self, a, b, *, allow_ints=True):
```

Comment:
We always "allow" ints. I think the name of this flag could make the behavior clearer.

Reply:
Suggestion? This flag controls whether the function can return integer dtypes.

Reply:
allow_int_comparisons?

Reply:
I was hoping a better name would help me more quickly understand the PR, so not sure if I have great suggestions yet. `allow_int_comparisons` still suggests to me that we'd throw an assertion error when trying to compare ints if false.

I think this is something to do with ensuring ints don't get converted to floats when unnecessary. I guess you would always want to return a floating dtype if you know you're going to need to use a floating tolerance... maybe something along the lines of `has_floating_tolerance`? I'd want it to be clear to the reader why an extra boolean is required. I think most readers will expect to be able to get the right comparison dtype from the dtypes of the parameters.

Maybe it should also be `comparison_dtype` instead of `compare_type`? The verb form suggests to me that this is asserting that dtypes match, but that doesn't seem to be what this does.

Reply:
New name sounds better. I removed the allow_ints kwarg to simplify the code.